midend/hsIndexSimplify.h

```diff
@@ -0,0 +1,72 @@
+#ifndef _MIDEND_HSINDEXSIMPLIFY_H_
+#define _MIDEND_HSINDEXSIMPLIFY_H_
+
+#include "frontends/common/resolveReferences/referenceMap.h"
+#include "frontends/p4/typeChecking/typeChecker.h"
+#include "frontends/p4/typeMap.h"
+#include "ir/ir.h"
+
+namespace P4 {
+
+/// This class finds a header stack with non-concrete index 
+/// which should be eliminated in unfolding if @a isFinder is true.
+/// Otherwise it substitutes index of a header stack in all occuarence of @a arrayIndex.
+class HSIndexFindOrTransform : public Transform {
```

Comment:
I see no benefit for this class to do double duty. Why don't you break it into two classes, an Inspector and a Transform.
They can communicate via a struct that keeps the fields that need to be used in common. This pattern is used a lot in the compiler.